src/vs/workbench/services/workspaces/common/workspaceTrust.ts

```diff
@@ -70,33 +104,78 @@ export class WorkspaceTrustManagementService extends Disposable implements IWork
-		@IWorkbenchEnvironmentService protected readonly environmentService: IWorkbenchEnvironmentService,
-		@IWorkspaceContextService protected readonly workspaceService: IWorkspaceContextService
+		@IWorkbenchEnvironmentService private readonly environmentService: IWorkbenchEnvironmentService,
+		@IWorkspaceContextService private readonly workspaceService: IWorkspaceContextService,
+		@IRemoteAuthorityResolverService private readonly remoteAuthorityResolverService: IRemoteAuthorityResolverService,
+
 	) {
 		super();
 
+		this._canonicalWorkspace = this.workspaceService.getWorkspace();
+		this._initialized = false;
+		this._workspaceTrustInitializedPromise = new Promise((resolve) => {
+			this._workspaceTrustInitializedPromiseResolve = resolve;
+		});
+
 		this._trustState = new WorkspaceTrustState(this.storageService);
 		this._trustTransitionManager = this._register(new WorkspaceTrustTransitionManager());
 
 		this._trustStateInfo = this.loadTrustInfo();
+		this._trustState.isTrusted = this.calculateWorkspaceTrust();
 
 		this.registerListeners();
 	}
 
+	//#region private interface
+
 	private registerListeners(): void {
-		this._register(this.workspaceService.onDidChangeWorkspaceFolders(async () => await this.recalculateWorkspaceTrust()));
-		this._register(this.workspaceService.onDidChangeWorkbenchState(async () => await this.recalculateWorkspaceTrust()));
+
+		// Resolve the workspace uris and resolve the initialization promise
+		this.resolveCanonicalWorkspaceUris().then(async () => {
+			this._initialized = true;
+			await this.updateWorkspaceTrust();
+			this._workspaceTrustInitializedPromiseResolve();
+		});
+
+		// Remote - resolve remote authority
+		if (this.environmentService.remoteAuthority) {
+			this.remoteAuthorityResolverService.resolveAuthority(this.environmentService.remoteAuthority)
+				.then(async result => {
+					this._remoteAuthority = result;
+					await this.updateWorkspaceTrust();
+				});
+		}
+
+		this._register(this.workspaceService.onDidChangeWorkspaceFolders(async () => await this.updateWorkspaceTrust()));
+		this._register(this.workspaceService.onDidChangeWorkbenchState(async () => await this.updateWorkspaceTrust()));
 		this._register(this.storageService.onDidChangeValue(async changeEvent => {
 			/* This will only execute if storage was changed by a user action in a separate window */
 			if (changeEvent.key === this.storageKey && JSON.stringify(this._trustStateInfo) !== JSON.stringify(this.loadTrustInfo())) {
 				this._trustStateInfo = this.loadTrustInfo();
 				this._onDidChangeTrustedFolders.fire();
 
-				await this.recalculateWorkspaceTrust();
+				await this.updateWorkspaceTrust();
 			}
 		}));
 	}
 
+	private async getCanonicalUri(uri: URI): Promise<URI> {
+		return this.environmentService.remoteAuthority && uri.scheme === Schemas.vscodeRemote ?
+			await this.remoteAuthorityResolverService.getCanonicalURI(uri) : uri;
+	}
+
+	private async resolveCanonicalWorkspaceUris(): Promise<void> {
```

Comment:
This should also be called when the `onDidChangeWorkspaceFolders` and `onDidChangeWorkbenchState` events fire.

Reply:
those listeners are hooked up to call `updateWorkspaceTrust` without a trust state which re-calls this method to guarantee the latest state.